src/metrics.cpp

```diff
@@ -0,0 +1,166 @@
+/*****************************************************************************
+ *
+ * This file is part of Mapnik (c++ mapping toolkit)
+ *
+ * Copyright (C) 2017 Artem Pavlenko
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ *****************************************************************************/
+
+#ifdef MAPNIK_METRICS
+
+#include <mapnik/metrics.hpp>
+
+#include <mapnik/make_unique.hpp>
+
+#include <boost/next_prior.hpp>
+#include <sstream>
+#include <utility>
+
+namespace mapnik {
+
+const std::string measurement_str[TOTAL_ENUM_SIZE] =
+{
+    "UNKNOWN",
+    "Time (us)",
+    "Value",
+    "Calls"
+};
+
+measurement::measurement(int64_t value, measurement_t type)
+    : value_(value), type_(type)
+{
+}
+
+autochrono::autochrono(metrics* m, std::string name)
+    : metrics_(m), name_(std::move(name))
+{
+    start_ = clock_.now();
+}
+
+autochrono::~autochrono()
+{
+    auto ns = std::chrono::duration_cast<time_units>(clock_.now() - start_);
+    metrics_->add_t(name_, ns.count(), measurement_t::TIME_MICROSECONDS);
+}
+
+std::unique_ptr<autochrono> metrics::time(const std::string& name)
+{
+    return std::make_unique<autochrono>(this, name);
```

Comment:
To minimize overhead, I would not construct `autochrono` on the heap like this, but rather on the stack. I think this method is misleading.

Reply:
The idea was to have a simple method to calculate time metrics. I might be wrong, but considering the current use in the feature style processor [measure_time](https://github.com/Algunenano/mapnik/blob/a24a6d50a9d6642c0cf8ef410a35860971e87367/include/mapnik/feature_style_processor_impl.hpp#L693) it should be movable all the way up and the overhead to be minimal.

Also, it's difficult to handle optional objects in the stack: 
```c++
if (metrics_enabled_)
{
     autochrono t(metrics_, name);
}
```
This would not work as it would destroy the scope of t. On the other hand you could use boost::optional which I think would have the same overhead, or handle passing nullptr to autochrono but *I think* that would slightly increase the overhead when you are not using metrics.
So I could switch it the calls to:
```c++
autochrono t(metrics_enabled_ ? metrics_ : nullptr, name)
```
What do you think?

Reply:
I think it's fine then. I was thinking why should be `metrics` dependent on `autochrono`.